File: Problem_Solving/leetcode/Sequential(Non-Linear)_data_structure/Tree/226_Invert_Binary_Tree/invertTree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# 파이썬 방식
def invertTree(self, root: TreeNode) -> TreeNode:
    # 재귀 방식인데 맨 밑부터 백트래킹해가며 스왑하는 Bottom-Up 방식
    if root:
        root.left, root.right = \
            invertTree(root.right), invertTree(root.left)
        return root
    return None

File: Problem_Solving/leetcode/Sequential(Non-Linear)_data_structure/Tree/226_Invert_Binary_Tree/invertTree.cpp
/*
    226. Invert Binary Tree
        Given the root of a binary tree, invert the tree,
        and return its root.



    Example 1:
        Input: root = [4,2,7,1,3,6,9]
        Output: [4,7,2,9,6,3,1]

    Example 2:
        Input: root = [2,1,3]
        Output: [2,3,1]

    Example 3:
        Input: root = []
        Output: []

    Constraints:
        The number of nodes in the tree is in the range [0, 100].
        -100 <= Node.val <= 100
*/
#include <stack>

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    TreeNode *invertTree(TreeNode *root)
    {
        std::stack<TreeNode*> stack;
        stack.push(root);

        while (stack.size())
        {
            TreeNode* node = stack.top();
            stack.pop();

            if (node)
            {
                TreeNode* temp = node->left;
                node->left = node->right;
                node->right = temp;
                
                stack.push(node->left);
                stack.push(node->right);
            }
        }

        return root;
    }
};